### PR TITLE
Crawler: try GET if HEAD raises 400 and stream GET requests

### PR DESCRIPTION
**Pull request by abulte**
```
>>> import requests
>>> requests.head('https://echanges.dila.gouv.fr/OPENDATA/DTD_LEGIFRANCE')
<Response [400]>
>>> requests.get('https://echanges.dila.gouv.fr/OPENDATA/DTD_LEGIFRANCE')
<Response [200]>
```

### CODE REVIEW

**Comment by jdesboeufs** [comment]
Maybe a by-domain opt-in should be better.
Too agressive for real Bad Request…

**Comment by abulte** [comment]
@oncletom Yes, quite a few servers do not care about HEAD requests and will just say the URL does not exist.

@jdesboeufs I see your point, I'm not sure it's worth the hassle though. Maintaining the list... And I'm worried about all the offending servers we won't catch (have to spot them on datagouv manually).

**Comment by abulte** [comment]
FIY, the URLs currently with `400` status in production:

```
http://echanges.bnf.fr/PIVOT/databnf_all_rdf_xml.tar.gz?user=databnf&password=databnf
http://clc.developpement-durable.gouv.fr/geoserver/gwc/service/wmts?
http://clc.developpement-durable.gouv.fr/geoserver/gwc/service/wms?
http://data.lacub.fr/wfs?key=BNHSR7SOSO&layer=CI_PARK_P&request=GetCapabilities&service=wms
http://data.lacub.fr/wfs?key=BNHSR7SOSO&layer=CI_PARK_P&request=GetCapabilities&service=wfs
http://echanges.bnf.fr/RELIURES/dump_reliures_xml_20130411.tar.gz?user=reliures&password=reliures2013
http://www.bnf.fr/fr/collections_et_services/livre_presse_medias/a.archives_internet.html
http://www.bnf.fr/fr/professionnels/depot_legal_definition/i.depot_legal_loi/s.depot_legal_loi_code.html?first_Rub=non#SHDC__Attribute_BlocArticle0BnF
http://data.lacub.fr/wms?key=
http://www.bnf.fr/documents/dl_obs_indicateurs_musique_2011.xls
http://www.bnf.fr/documents/dl_obs_audiovisuel_2013.xls
http://www.bnf.fr/documents/dl_obs_indicateurs_livres_2011.xls
http://www.bnf.fr/documents/dl_observatoire_p%C3%A9riodiques_2012.xls
http://www.bnf.fr/documents/dl_obs_indicateurs_doc_carto_2011.xls
http://www.bnf.fr/documents/dl_obs_livres_2014.xls
http://www.bnf.fr/documents/dl_obs_indicateurs_sites_web_2011.xls
http://www.bnf.fr/documents/dl_obs_indicateurs_estampes_2011.xls
http://www.bnf.fr/documents/dl_obs_musique_2013.xls
http://www.bnf.fr/documents/dl_observatoire_2012.pdf
http://www.bnf.fr/documents/dl_obs_sites_web_2013.xls
http://www.bnf.fr/documents/dl_observatoire_periodiques_2013.xls
http://www.bnf.fr/documents/dl_obs_audiovisuel_2014.xls
http://www.bnf.fr/documents/dl_observatoire_livres_2012.xls
http://www.bnf.fr/documents/dl_obs_musique_2014.xls
http://www.bnf.fr/documents/dl_observatoire_estampes_2012.xls
http://www.bnf.fr/documents/dl_observatoire_sites_web_2012.xls
http://www.bnf.fr/documents/dl_obs_doc_carto_2014.xls
http://www.bnf.fr/documents/dl_observatoire_doc_carto_2012.xls
http://www.bnf.fr/documents/dl_obs_periodiques_2014.xls
http://www.bnf.fr/documents/dl_obs_indicateurs_audiovisuel_2011.xls
http://www.bnf.fr/documents/dl_obs_doc_carto_2013.xls
http://www.bnf.fr/documents/dl_obs_estampes_2013.xls
http://www.bnf.fr/documents/dl_obs_sites_web_2014.xls
http://www.bnf.fr/documents/dl_obs_estampes_2014.xls
http://www.bnf.fr/documents/dl_obs_periodiques_2011.xls
http://www.bnf.fr/documents/dl_observatoire_audiovisuel_2012.xls
http://www.bnf.fr/documents/dl_observatoire_musique_2012.xls
http://www.bnf.fr/documents/dl_observatoire_livres_2013.xls
http://www.bnf.fr/fr/professionnels/depot_legal_definition/s.depot_legal_observatoire.html
http://www.bnf.fr/documents/dl_observatoire_2011.pdf
http://www.bnf.fr/fr/professionnels/donnees_bnf_recuperables/a.notices_marc.html
http://www.bnf.fr/fr/professionnels/donnees_bnf_recuperables/a.donnees_rdf.html
http://www.bnf.fr/fr/professionnels/donnees_bnf_recuperables/a.notices_biblio_dublincore.html
http://www.bnf.fr/fr/professionnels/donnees_bnf_recuperables/a.notices_gallica.html
http://geo.brest-metropole.fr/arcgis/services/public/GPB_DEP/MapServer/WFSServer?service=wfs%26version=1.0.0%26request=GetFeature%26typename=public_GPB_DEP:TRANSPORT_COLLECTIF_-_Arr%C3%AAts_de_Bus_et_de_Tramway
http://geo.brest-metropole.fr/arcgis/services/public/GPB_DEP/MapServer/WFSServer?service=wfs%26version=1.0.0%26request=GetFeature%26typename=public_GPB_DEP:TRANSPORT_COLLECTIF_-_Lignes_de_Bus_et_de_Tramway
http://echanges.dila.gouv.fr/anonymous/OPENDATA/REF_ANNUAIRE/SCHEMA/modele_dila_bdi_toPublish.zip
http://echanges.dila.gouv.fr/anonymous/OPENDATA/REF_ANNUAIRE/2017-COURANT/export_data_gouv_20170328.zip
http://echanges.dila.gouv.fr/anonymous/OPENDATA/REF_ANNUAIRE/2017-COURANT/export_data_gouv_latest.zip
http://echanges.dila.gouv.fr/OPENDATA/REF_ANNUAIRE/2017-COURANT/export_data_gouv_latest.zip
https://echanges.dila.gouv.fr/OPENDATA/RefOrgaAdminEtat/FluxHistorique/2017-FluxCourant/dila_refOrga_admin_Etat_fr_latest.zip
https://echanges.dila.gouv.fr/OPENDATA/RefOrgaAdminEtat/Modele/modele_dila_bdi_toPublish.zip
https://echanges.dila.gouv.fr/OPENDATA/Liste_Greffes_BODACC/DILA_Liste_greffes_bodacc_20170213.xlsx
https://echanges.dila.gouv.fr/OPENDATA/Protocole_des_JAAI/DILA_Protocole_JAAI__20170510.xml
http://data.lacub.fr/wfs?key=WWFUHCRKSZ&&layers=en_empac_p
http://data.lacub.fr/wms?key=WWFUHCRKSZ&&layers=en_frcol_s
http://data.lacub.fr/wms?key=WWFUHCRKSZ&&layers=en_empac_p
https://echanges.dila.gouv.fr/OPENDATA/BOAMP/2017/BOAMP-J-AO_2017_001001.taz
https://echanges.dila.gouv.fr/OPENDATA/BOAMP/2017/BOAMP-J-AO_2017_008004.taz
https://echanges.dila.gouv.fr/OPENDATA/Protocole_des_JAAI/DILA_ProtocoleJAAI_Presentation20170510.pdf
https://echanges.dila.gouv.fr/OPENDATA/ASSOCIATIONS/
https://echanges.dila.gouv.fr/OPENDATA/ASSOCIATIONS/Documentation/
https://echanges.dila.gouv.fr/OPENDATA/JORF/
https://echanges.dila.gouv.fr/OPENDATA/DTD_LEGIFRANCE
https://echanges.dila.gouv.fr/OPENDATA/JORFSIMPLE/
https://echanges.dila.gouv.fr/OPENDATA/BODACC/
https://echanges.dila.gouv.fr/OPENDATA/JADE/
https://echanges.dila.gouv.fr/OPENDATA/CAPP/
https://echanges.dila.gouv.fr/OPENDATA/CAPP/DTD+JURI+FREEMIUM/
https://echanges.dila.gouv.fr/OPENDATA/DTD_LEGIFRANCE/
https://echanges.dila.gouv.fr/OPENDATA/LEGI/
https://echanges.dila.gouv.fr/OPENDATA/CASS/
https://echanges.dila.gouv.fr/OPENDATA/BALO/
https://echanges.dila.gouv.fr/OPENDATA/INCA/
https://echanges.dila.gouv.fr/OPENDATA/KALI/
https://echanges.dila.gouv.fr/OPENDATA/COMPTES_DES_ASSOCIATIONS/
https://echanges.dila.gouv.fr/OPENDATA/COMPTES_DES_ASSOCIATIONS/DOCUMENTATION/
https://echanges.dila.gouv.fr/OPENDATA/SARDE/
https://echanges.dila.gouv.fr/OPENDATA/SARDE/DTD+SARDE+FREEMIUM/
https://echanges.dila.gouv.fr/OPENDATA/CIRCULAIRES/
https://echanges.dila.gouv.fr/OPENDATA/INCA/DTD+JURI+FREEMIUM/
https://digital.iservices.rte-france.com/open_api/wholesale_market/v1/epex_spot_power_exchanges
https://echanges.dila.gouv.fr/OPENDATA/CNIL/
https://echanges.dila.gouv.fr/OPENDATA/CONSTIT/
https://echanges.dila.gouv.fr/OPENDATA/CONSTIT/DTD+CONSTIT+FREEMIUM/
https://www.ideobfc.fr/geonetwork/srv/api/records/0a78e68d-738c-4284-ae64-e02e74859055/attachments/beta_compositions_communales_EPCI_2017_ARD.xlsx
https://echanges.dila.gouv.fr/OPENDATA/Protocole_du_Gouvernement/Gouvernement_protocole_20170621.xml
https://echanges.dila.gouv.fr/OPENDATA/Protocole_du_Gouvernement/DILA_ProtocoleGouvernement_Presentation20170510.pdf
https://www.ideobfc.fr/geonetwork/srv/api/records/544a9508-542d-4a21-a4ca-d4f09931be9f/attachments/zppa_meta.odt
https://echanges.dila.gouv.fr/OPENDATA/RefOrgaAdminEtat/Documentation/DILA_RefOrgaAdminEtat_Presentation_20170421.pdf
https://echanges.dila.gouv.fr/OPENDATA/LEGI/DILA_LEGI_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/KALI/DILA_KALI_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/JORF/DILA_JORF_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/CAPP/DILA_CAPP_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/CASS/DILA_CASS_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/INCA/DILA_INCA_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/JADE/DILA_JADE_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/CIRCULAIRES/DILA_CIRCULAIRES_Presentation_20170828.pdf
https://echanges.dila.gouv.fr/OPENDATA/SARDE/DILA_SARDE_Presentation_20170830.pdf
https://echanges.dila.gouv.fr/OPENDATA/Protocole_du_Gouvernement/DILA_Protocole-Gouvernement_Presentation20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/Protocole_des_JAAI/DILA_Protocole-JAAI_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/BALO/DILA_BALO_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/BOAMP/DILA_BOAMP_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/BOAMP/2017/
https://echanges.dila.gouv.fr/OPENDATA/BOAMP/DILA_API-BOAMP_Presentation_20170831.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-choix-forfait-BOAMP_Presentation_20170828.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-choix-de-formulaire-BOAMP_Presentation_20170828.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-calcul-interets-moratoires_Presentation_20170829.pdf
https://echanges.dila.gouv.fr/OPENDATA/BODACC/2017/
https://echanges.dila.gouv.fr/OPENDATA/BODACC/DILA_BODACC_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/Liste_Greffes_BODACC/DILA_Liste-greffes_BODACC_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-cout-certificat-immatriculation_Presentation_20170828.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-calcul-pension-alimentaire_Presentation_20170829.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-calcul-frais-de-succession_Presentation_20170828.pdf
https://echanges.dila.gouv.fr/OPENDATA/DILA_Simulateur-calcul-gratification-minimale-stagiaire_Presentation_20170829.pdf
https://echanges.dila.gouv.fr/OPENDATA/ASSOCIATIONS/ASS_2017/
https://echanges.dila.gouv.fr/OPENDATA/ASSOCIATIONS/DILA_ASSOCIATIONS_Presentation_20170824.pdf
https://echanges.dila.gouv.fr/OPENDATA/COMPTES_DES_ASSOCIATIONS/DILA_COMPTES-ASSOCIATIONS_Presentation_20170824.pdf
```

**Comment by abulte** [comment]
@davidbgk @noirbizarre @odtvince An opinion on this? ⬆️

**Comment by abulte** [comment]
I added `stream=True` on the `GET` requests: this will prevent from downloading the whole files/page. Less of a strain on local and distant servers.

**Comment by abulte** [comment]
But I need to close the connections too, incoming.

**Comment by jdesboeufs** [comment]
2 identified domains + OGC services (WMS, WFS, WMTS) which would rely on our INSPIRE gateway.

**Comment by abulte** [comment]
@jdesboeufs And the other ones seems to be real bad requests... Which advocates for your solution, I agree.

**Comment by abulte** [comment]
@jdesboeufs Done. Do you think I should put the OGC services in the list for now? On a side note, would a streaming GET request trigger computations on those services?

NB: I'll put the list in a config file and write better tests on the `dev` branch where I have everything I need for that.

**Comment by oncletom** [comment]
@abulte can the list of offenders be set by udata + udata-croquemort plugin? It would keep this package less opinionated with these domains.

I'm still doubtful about having the need of an additional configuration element for just 2 entries.

**Comment by abulte** [comment]
@oncletom yes you're right, in the future architecture those domains should probably be maintained on the `udata-croquemort` side and `croquemort` just receive an argument wether to use HEAD method or not.

**Comment by abulte** [comment]
@oncletom Still doubtful too :-/

File: croquemort/crawler.py [hunk 1]
```diff
@@ -45,17 +51,21 @@ def check_url(self, url_group_frequency):
             if frequency:
                 self.storage.store_frequency(url, group, frequency)
         try:
-            try:
-                response = session.head(url, allow_redirects=True,
-                                        timeout=HEAD_TIMEOUT)
-            except requests.exceptions.ReadTimeout:
-                # simulate 404 to trigger GET request below
-                response = FakeResponse(status_code=404, headers={})
+            head_offend = any([re.match('^(http|https|ftp)://%s' % d, url)
```
Comment (noirbizarre) [inline croquemort/crawler.py]: Maybe faster to extract the domain once with `urlsplit()` and test if it's in the list, no ?

File: croquemort/crawler.py [hunk 2]
```diff
@@ -11,6 +11,12 @@
 
 HEAD_TIMEOUT = 10  # in seconds
 GET_TIMEOUT = 3 * 60  # in seconds
+# domains that won't answer correctly to HEAD requests
+# TODO move to config file
+KNOWN_HEAD_OFFENDER_DOMAINS = [
```
Comment (noirbizarre) [inline croquemort/crawler.py]: :+1: but would be awesome to have this externalisable (ie. not an hardcoded list but a configurable entry) so we may keep this up to date without having to perform a release (just add the new domains to ansible vars and just redeploy the conf)

File: tests/test_integrations.py [hunk 3]
```diff
@@ -237,6 +237,23 @@ def test_crawling_url(runner_factory, rpc_proxy_factory):
     assert storage.store_metadata.call_count == 1
 
 
+# TODO use config file for KNOWN_HEAD_OFFENDER_DOMAINS (v2)
+# TODO use request mock to check HEAD is not called (v2)
```
Comment (noirbizarre) [inline tests/test_integrations.py]: :+1: for the TODOs

File: croquemort/crawler.py (hunk 2 above)
Comment (abulte) [reply croquemort/crawler.py]: Yep, like I said above I will put this in a config file in v2 which should arrive very shortly. I made some changes to the way the config is handled in dev and it's simpler if I make the change there.

File: croquemort/crawler.py (hunk 1 above)
Comment (abulte) [reply croquemort/crawler.py]: Maybe :-) I'll do that.

File: croquemort/crawler.py (hunk 2 above)
Comment (noirbizarre) [reply croquemort/crawler.py]: :+1:

**Comment by abulte** [comment]
@noirbizarre and do you think I should put the OGC services in the list too?

File: croquemort/crawler.py [hunk 4]
```diff
@@ -11,6 +12,13 @@
...
+    'www.bnf.fr',
```
Comment (davidbgk) [inline croquemort/crawler.py]: I think it should not be declared within croquemort but udata-croquemort?

Comment (abulte) [reply croquemort/crawler.py]: Probably a good idea, cf https://github.com/opendatateam/croquemort/pull/42#issuecomment-328051895.

**Pull request merged**